pythran/pythonic/numpy/random/choice.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef PYTHONIC_NUMPY_RANDOM_CHOICE_HPP
+#define PYTHONIC_NUMPY_RANDOM_CHOICE_HPP
+
+#include "pythonic/include/numpy/random/choice.hpp"
+#include "pythonic/include/numpy/random/generator.hpp"
+
+#include "pythonic/__builtin__/NotImplementedError.hpp"
+#include "pythonic/numpy/random/randint.hpp"
+#include "pythonic/types/ndarray.hpp"
+#include "pythonic/types/tuple.hpp"
+#include "pythonic/utils/proxy.hpp"
+
+#include <random>
+#include <algorithm>
+
+namespace pythonic
+{
+  namespace numpy
+  {
+    namespace random
+    {
+      /***********************************************************
+       * Implementation with long as first argument
+       **********************************************************/
+      template <size_t S, class P>
+      types::ndarray<long, S> choice(long max,
+                                     types::array<long, S> const& shape,
+                                     bool replace,
+                                     P const& p)
+      {
+        if(not replace)
+          throw pythonic::__builtin__::NotImplementedError(
+              "Choice without replacement is not implemented, ask if you want it");
+
+        types::ndarray<long, S> result{shape, types::none_type()};
+        std::discrete_distribution<long> distribution{p.begin(), p.end()};
+        std::generate(result.fbegin(), result.fend(),
+                      [&]() { return distribution(details::generator); });
+        return result;
+      }
+
+      template <class P>
+      types::ndarray<long, 1> choice(long max,
+                                     long size,
+                                     bool replace,
+                                     P && p)
+      {
+        return choice(max, types::array<long, 1>{{size}}, replace, std::forward<P>(p));
+      }
+
+      template <class T>
+      auto choice(long max, T&& size) -> decltype(randint(0, max, size))
+      {
+        return randint(0, max, size);
+      }
+
+      long choice(long max)
+      {
+        return randint(max);
+      }
+
+      /***********************************************************
+       * Implementation with array as first argument
+       **********************************************************/
+
+      template <class T>
+      typename T::dtype choice(T const& a)
```

Comment:
Well, it will not compile for any numpy expression (including ndarray) that's not 1D, because ``a.fast(i)`` is somehow equivalent to ``a[i]`` without bound check.